### PR TITLE
Change inventory endpoint to new format

### PR DESCRIPTION
**Pull request by Furdarius**
Inventory endpoint was changed to new format

Info about endpoints format: http://stackoverflow.com/a/17679641

Signed-off-by: Artemiy Ryabinkov <getlag@ya.ru>

### CODE REVIEW

**Comment by asamy** [comment]
Hi, thanks!

I got this with the new inventory example:

```
2017/02/04 00:53:27 main.go:26: Login successful
2017/02/04 00:53:28 main.go:35: -- AppID total asset count: 1
2017/02/04 00:53:28 main.go:37: -- Items on 620 2 (count 1)
2017/02/04 00:53:29 main.go:44: Item: WCC FLAG = 23308196
2017/02/04 00:53:29 main.go:35: -- AppID total asset count: 1
2017/02/04 00:53:29 main.go:37: -- Items on 4920 2 (count 1)
2017/02/04 00:53:29 main.go:44: Item: Eat Your Greens Shoulder Patch = 190647574130629115
2017/02/04 00:53:29 main.go:35: -- AppID total asset count: 1
2017/02/04 00:53:29 main.go:37: -- Items on 218620 2 (count 1)
2017/02/04 00:53:30 main.go:44: Item: Bodhi Safe = 236804570526174871
2017/02/04 00:53:30 main.go:35: -- AppID total asset count: 902
2017/02/04 00:53:30 main.go:37: -- Items on 730 2 (count 902)
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x63c680, 0xc42000c100)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/asamy/steam.(*Session).fetchInventory(0xc42007a640, 0x110000107110662, 0x2da, 0x2, 0x6835da, 0x7, 0x0, 0xc420127a10, 0xc42001a140, 0x2, ...)
	/home/asamy/gopkg/src/github.com/asamy/steam/inventory.go:176 +0xe1e
github.com/asamy/steam.(*Session).GetInventory(0xc42007a640, 0x110000107110662, 0x2da, 0x2, 0x6835da, 0x7, 0xc4202f08c0, 0x1, 0x1, 0x0, ...)
	/home/asamy/gopkg/src/github.com/asamy/steam/inventory.go:230 +0xb5
main.main()
	/home/asamy/gopkg/src/github.com/asamy/steam/examples/inventory/main.go:38 +0x6e4
exit status 2
```

File: inventory.go [hunk 1]
```diff
@@ -41,104 +76,167 @@ type InventoryAppStats struct {
...
-		MoreStart    interface{}     `json:"more_start"` // This can be a bool or a number...
-		Inventory    json.RawMessage `json:"rgInventory"`
-		Descriptions json.RawMessage `json:"rgDescriptions"`
-		/* Missing: rgCurrency  */
+		Assets              []Asset       `json:"assets"`
+		Descriptions        []Description `json:"descriptions"`
+		Success             int           `json:"success"`
+		HasMore             int           `json:"more_items"`
+		LastAssetID         string        `json:"last_assetid"`
+		TotalInventoryCount int           `json:"total_inventory_count"`
```
Comment (asamy) [inline inventory.go]: Since we have TotalInventoryCount here, why don't we use it for *items pre-allocation?

File: inventory.go [hunk 2]
```diff
@@ -41,104 +76,167 @@ type InventoryAppStats struct {
...
+		tmpKey := fmt.Sprintf("%d_%d", asset.ClassID, asset.InstanceID)
+		if tmpKey != lastKey {
+			lastKey = tmpKey
+
+			descPos++
+		}
+
+		picts := ItemPict{
+			Standart: fmt.Sprintf(PicturePathPattern, response.Descriptions[descPos].IconURL),
+		}
```
Comment (asamy) [inline inventory.go]: This seems to be faulty in my test, the items run out of descriptions.

File: inventory.go [hunk 3]
```diff
@@ -4,22 +4,57 @@ import (
...
+	LangEng = "english"
+	LangRus = "russian"
+)
+
+const (
+	PicturePathPattern = "http://steamcommunity-a.akamaihd.net/economy/image/%s"
+)
+
+type ItemPict struct {
+	Standart string `json:"standart"`
```
Comment (asamy) [inline inventory.go]: Typo: "standard"

File: inventory.go [hunk 4]
```diff
@@ -41,104 +76,167 @@ type InventoryAppStats struct {
...
+	//   C        2
+	//   C        2
+	//   D        3
+	descPos := 0
+	lastKey := fmt.Sprintf("%d_%d", response.Assets[0].ClassID, response.Assets[0].InstanceID)
+	for _, asset := range response.Assets {
+		tmpKey := fmt.Sprintf("%d_%d", asset.ClassID, asset.InstanceID)
+		if tmpKey != lastKey {
+			lastKey = tmpKey
+
```
Comment (asamy) [inline inventory.go]: How about renaming the _ to i then checking i != 0?

File: inventory.go [hunk 5]
```diff
@@ -41,104 +76,167 @@ type InventoryAppStats struct {
...
-	case int:
-		return uint32(v), nil
-	case uint:
-		return uint32(v), nil
-	case bool:
-		break
-	default:
-		return 0, fmt.Errorf("parseInventory: missing implementation for type %v", v)
+	hasMore = response.HasMore == 1
+
```
Comment (asamy) [inline inventory.go]: != 0 better?

File: inventory.go [hunk 6]
```diff
@@ -4,22 +4,57 @@ import (
...
+type Lang string
+
+const (
+	LangEng = "english"
+	LangRus = "russian"
+)
+
+const (
+	PicturePathPattern = "http://steamcommunity-a.akamaihd.net/economy/image/%s"
+)
```
Comment (asamy) [inline inventory.go]: Add the endpoint URL here, too.

File: inventory.go [hunk 7]
```diff
@@ -41,104 +76,167 @@ type InventoryAppStats struct {
...
+	descPos := 0
+	lastKey := fmt.Sprintf("%d_%d", response.Assets[0].ClassID, response.Assets[0].InstanceID)
+	for _, asset := range response.Assets {
+		tmpKey := fmt.Sprintf("%d_%d", asset.ClassID, asset.InstanceID)
+		if tmpKey != lastKey {
+			lastKey = tmpKey
+
+			descPos++
+		}
+
```
Comment (asamy) [inline inventory.go]: Perhaps to fix the panic I pointed out is to check if descPos+1 < len(descriptions) and move the picture initialization here?

**Comment by Furdarius** [comment]
"index out of range" error fixed. 

I wasn't right about assets and descriptions order, so now done it with classical method (Map, where key is asset unique identifier and value is position in descriptions array)

File: inventory.go (hunk 1 above)
Comment (Furdarius) [inline inventory.go]: @asamy We can do it, but not with current realization. Now `items` array initialized in `GetInventory` method, but filled in `fetchInventory`, so on initialization time we don't know totalCount

**Comment by Furdarius** [comment]
By the way, where are you taking `steamSharedSecret`? Is it codes from steam app (e.g. `G96NH`, etc...)?

**Comment by asamy** [comment]
No, it's the shared secret generated by initiating two factor on an account, you either have to do it manually (use twofactor.go api) or AFAIK root your device to get it, then you can use it to generate codes like the steam app with the totp.go api.

Since this API is too fast, I sometimes get "Called method busy, action not taken (10)" error after 2 consecutive requests (less than a sec) in reality this wouldn't really be needed since steam is a tortoise in general, so it's okay, so I have added a small delay in inventory example.

Thanks for your effort!

**Pull request merged**